Esxi driver

b0r6 opened the pull request:
The executor needs a ssh-key to be present in `/etc/openvdc/ssh/esxi` to be able to access vmkfstools on the esxi-machine, which it uses to clone vmdks. For example, running openvdc run centos/7/esxi clones a base CentOS7 vm into one called i-000000001

The executor waits for the instance vm to get an ip.  When it has gotten one, the executor sets the state to RUNNING. This was a simple way for the executor to make sure that the instance was actually up and running, but the downside is that it requires an ip to be set, so at some later point it will probably use a different method for determining this.
`openvdc run centos/7/esxi '{"interfaces":[{"Ipv4Addr":"10.0.100.80"}]}'`

b0r6 committed:
Added script that builds a machine in esxi, installs some required stuff and then shuts it down.

b0r6 committed:
Added script that builds a machine in esxi, installs some required stuff and then shuts it down.

b0r6 committed:
Output vm.create command

b0r6 committed:
Output vm.create command

b0r6 committed:
Shut down VM after installing yum packages

b0r6 committed:
Shut down VM after installing yum packages

b0r6 committed:
Attempt to copy prepared vm & fixed command

b0r6 committed:
Attempt to copy prepared vm & fixed command

b0r6 committed:
http->vi

b0r6 committed:
http->vi

b0r6 committed:
Make sure to not export attached iso file

b0r6 committed:
Make sure to not export attached iso file

b0r6 committed:
Use backup to create VM if rebuild = false. Also set VMName to be the same as BranchName

b0r6 committed:
Use backup to create VM if rebuild = false. Also set VMName to be the same as BranchName

b0r6 committed:
Re-arrange and fix some stuff

b0r6 committed:
Re-arrange and fix some stuff

b0r6 committed:
Check for ovftool

b0r6 committed:
Check for ovftool

b0r6 committed:
Install openvdc

b0r6 committed:
Install openvdc

b0r6 committed:
Enable scheduler

b0r6 committed:
Enable scheduler

b0r6 committed:
disable stuff during installation

b0r6 committed:
disable stuff during installation

b0r6 committed:
Change IP

b0r6 committed:
Change IP

b0r6 committed:
Get VM IP when restoring backup

b0r6 committed:
Get VM IP when restoring backup

b0r6 committed:
Added ssh_cmd

b0r6 committed:
Added ssh_cmd

b0r6 committed:
Added yum_install function and fixed some bugs

b0r6 committed:
Added yum_install function and fixed some bugs

b0r6 committed:
build vms on datastore2

b0r6 committed:
build vms on datastore2

b0r6 committed:
Some adjustments

b0r6 committed:
Some adjustments

b0r6 committed:
Added basic vmware template

b0r6 committed:
Updated proto files

b0r6 committed:
Added attributes.vmware

b0r6 committed:
Scheduler: accept vmware

b0r6 committed:
Something went wrong last time, tried generating these files again

b0r6 committed:
Added base for hypervisor/vmware

b0r6 committed:
Added some esxi config options

b0r6 committed:
Added vmware

b0r6 committed:
Updated these again

b0r6 committed:
Fixed a mistake

b0r6 committed:
Added vmware driver

b0r6 committed:
Updated to include vmware

b0r6 committed:
Added vmware

b0r6 committed:
added handlers/vm/vmware

b0r6 committed:
added vmware.json

b0r6 committed:
vmware->esxi

b0r6 committed:
Set esxiUrl

b0r6 committed:
Create new client

b0r6 committed:
Rename these as well

b0r6 committed:
import handlers/vm/esxi

b0r6 committed:
Added esxi driver

b0r6 committed:
Updated vendor.json

b0r6 committed:
GetString->GetBool

b0r6 committed:
Fixed invalid variable name

b0r6 committed:
Build url after loading config

b0r6 committed:
fix url

b0r6 committed:
Create new folder in Datastore

b0r6 committed:
Added config option for esxihost ssh-key

b0r6 committed:
ssh into esxi-host and clone vmdk

b0r6 committed:
Copy .vmx file

b0r6 committed:
Register new VM

b0r6 committed:
Rename VM to instance-id

b0r6 committed:
Boot up new VM

b0r6 committed:
Added EsxiCmd function to avoid duplicated code

b0r6 committed:
gofmt

b0r6 committed:
Added destroy command

b0r6 committed:
Added start & stop

b0r6 committed:
Added config options

b0r6 committed:
Replace hardcoded value with configurable one

b0r6 committed:
Added reboot

b0r6 committed:
Made some hardcoded stuff configurable

b0r6 committed:
Added script for setting esxi vm ip

b0r6 committed:
Set Ipv4, upload and run script

b0r6 committed:
Restart network

b0r6 committed:
Use scriptpath from config file

b0r6 committed:
Merged in master branch

b0r6 committed:
Fixed bug

b0r6 committed:
Update cluster.pb.go

b0r6 committed:
govendor.json: fixed govc packages

b0r6 committed:
See if reverting this prevents error

b0r6 committed:
Added esxi_test.go

b0r6 committed:
Fixed bug that returned empty template

b0r6 committed:
Use env variables

b0r6 committed:
Rename some things

b0r6 committed:
Added esxi node to Jenkinsfile

b0r6 committed:
Expect env file

b0r6 committed:
set -ex -o pipefail

b0r6 committed:
Attempting to add esxi node env variables

b0r6 committed:
Rename stage

b0r6 committed:
Disable these steps for now

b0r6 committed:
Forgot that this was groovy. # -> //

b0r6 committed:
Revert check, add it at some later point if necessary

b0r6 committed:
Set this in build script for now

b0r6 committed:
Adjustments to build script

b0r6 committed:
Update env var

b0r6 committed:
Disable this step for now

b0r6 committed:
On second thought, don't.

b0r6 committed:
Prevent infinite loop

b0r6 committed:
Shutdown and remove VM afterwards

b0r6 committed:
Increase sleep time

b0r6 committed:
Fixed issue where the wrong command was being used

b0r6 committed:
Remove check for two variables that aren't used anymore.

b0r6 committed:
When having detected a vm ip; sleep for a moment before trying to do anything.

b0r6 committed:
Add openvdc-executor to bin

b0r6 committed:
Merged ci-esxi into esxi-driver

b0r6 committed:
Re-add these steps

b0r6 committed:
Fixed typo

b0r6 committed:
Merge branch 'ci-esxi' into esxi-driver

b0r6 committed:
Remove some flags

b0r6 committed:
Merge branch 'master' into ci-esxi

b0r6 committed:
Try to skip echoing a few lines

b0r6 committed:
Test

b0r6 committed:
Try avoid outputting some vars

b0r6 committed:
Another test

toros11 committed:
Use ssh session to perform ssh command

toros11 committed:
Merge remote-tracking branch 'remotes/origin/ci-esxi' into feature-esxi-console

b0r6 committed:
Re-enable steps

toros11 committed:
Merge branch 'ci-esxi' into feature-esxi-console

toros11 committed:
Created interface methods for implementation of basic console feature

b0r6 committed:
Merge branch 'ci-esxi' into esxi-driver

b0r6 committed:
Check if interface is set

b0r6 committed:
Set variable

b0r6 committed:
Added esxi_test.go

toros11 committed:
Create serial port for attaching tty to

toros11 committed:
Connect/Disconnect the serial port on start/stop

toros11 committed:
Update vendors

toros11 committed:
Attach console to serial port of the instance

b0r6 committed:
Merge branch 'master' into ci-esxi

b0r6 committed:
Merge branch 'feature-esxi-console' into esxi-driver

b0r6 committed:
Merge branch 'master' into esxi-driver

b0r6 committed:
Setup config file

b0r6 committed:
Fixed a line where an old function was being called

toros11 committed:
Temprorarily disable this function call in order to not generate error

toros11 committed:
Merge branch 'master' into feature-esxi-console

Conflicts:
	model/model.pb.go
	registry/bindata_assetfs.go

b0r6 committed:
set up ssh-key that's currently needed by the executor

b0r6 committed:
Merge branch 'master' into ci-esxi

b0r6 committed:
Fixed invalid character

b0r6 committed:
added cleanup

toros11 committed:
Refactor serial console

Created utility function that connects to a serial port and attaches
the file descriptors of the ssh console. The expected i/o of the
serial console doesn't change between the hypervisors and implementing
it for each driver caused unnecessary code duplication.

toros11 committed:
Stick to naming conventions

b0r6 committed:
Merge branch 'ci-esxi' into esxi-driver

b0r6 committed:
Added executor-esxi package

b0r6 committed:
Commented out some lines for now for debugging purposes.

b0r6 committed:
Some adjustments

b0r6 committed:
Fix datastore.download call

toros11 committed:
Send new line to trigger the console

b0r6 committed:
Add vars

b0r6 committed:
re-enable cleanup and ci steps

b0r6 committed:
Merge branch 'ci-esxi' into esxi-driver

b0r6 committed:
Added basic test

b0r6 committed:
Added acceptance-test-esxi

b0r6 committed:
Add acceptance-test-esxi to build

b0r6 committed:
Disable these while debugging

b0r6 committed:
Attempt to run test

b0r6 committed:
Install test

b0r6 committed:
Fix invalid package name

b0r6 committed:
Install missing package

b0r6 committed:
Change ssh-key folder

b0r6 committed:
Re-enable steps

toros11 committed:
Add remote command execution to esxi console.

This feature depends on the custom vmware toolbox agent being
installed on the guest.

toros11 committed:
Don't try to close nil connection

toros11 committed:
Fix error return struct

toros11 committed:
Return the exit code

unakatsuo commented on hypervisor/esxi/esxi.go:
```diff
@@ -0,0 +1,309 @@
+		a = append(a, args[i])
+	}
+
+	cli.Run(a)
```
How can we handle errors from ``cli.Run``? All the lines where ``esxiCmd()`` is called are missing to handle error.

toros11 replied:
The `cli.Run()` function returns an int, `0` on success, `1` on failure. The error message is printed directly to os.Stderr by the function itself. 

The options here seems to be to either return an error from the `esxiCmd()` if the return code was `1`, 
or return the function call itself and let the calling function return the error based on the return code

unakatsuo commented on hypervisor/esxi/esxi.go:
```diff
@@ -0,0 +1,309 @@
+	a = append(a, args[0])
+	a = append(a, fmt.Sprintf("-dc=%s", settings.EsxiDatacenter))
+	a = append(a, fmt.Sprintf("-k=%s", strconv.FormatBool(settings.EsxiInsecure)))
+	a = append(a, fmt.Sprintf("-u=%s", settings.EsxiUrl))
```
Here can be rewrite:

```
a := []string {
args[0],
fmt.Sprintf("-dc=%s", settings.EsxiDatacenter),
fmt.Sprintf("-k=%s", strconv.FormatBool(settings.EsxiInsecure)),
fmt.Sprintf("-u=%s", settings.EsxiUrl),
}
```

unakatsuo commented on hypervisor/esxi/esxi_console.go:
```diff
@@ -0,0 +1,145 @@
+				}
+			}
+		}
+		defer waitClosed.Done()
```
``defer`` should come to the beginning of the ``go func()``.

unakatsuo commented on hypervisor/esxi/esxi_console.go:
```diff
@@ -0,0 +1,145 @@
+				}
+			}
+		}
+		defer waitClosed.Done()
```
``defer`` should come to the beginning of the ``go func()``.

unakatsuo commented on hypervisor/esxi/esxi.go:
```diff
@@ -0,0 +1,309 @@
+	settings.EsxiHostSshkey = sub.GetString("hypervisor.esxi-host-sshkey")
+	settings.EsxiVmUser = sub.GetString("hypervisor.esxi-vm-user")
+	settings.EsxiVmPass = sub.GetString("hypervisor.esxi-vm-pass")
+	settings.EsxiVmDatastore = sub.GetString("hypervisor.esxi-vm-datastore")
```
Should validate the required parameters and then return error if they were blank.

unakatsuo commented on hypervisor/esxi/esxi.go:
```diff
@@ -0,0 +1,309 @@
+
+	esxiInfo := fmt.Sprintf("%s:%s@%s", settings.EsxiUser, settings.EsxiPass, settings.EsxiIp)
+	u, _ := url.Parse("https://")
+	u.Path = path.Join(u.Path, esxiInfo, "sdk")
```
Let ``url.Parse()`` validate the syntax and return the format error.

```
u, err := url.Parse("https://" + esxiInfo + "/sdk")
if err != nil {
  return errors.Wrap(err, "Failed to parse url for ESXi server")
}
```

unakatsuo reviewed (changes requested):
Please add error check to all ``esxiCmd()`` calls and remove ``// TODO: error checks`` comments.

toros11 committed:
create service tty on serial port if it does not exist

toros11 committed:
Temporarily disable waiting for ip

toros11 committed:
Use channel to store bufferd output

toros11 committed:
Refactored esxi driver

Made the api calling wrapper function return error when the
return code from API indicates failure and allow batch exection
of cli commands.

toros11 committed:
Validate esxi server url

toros11 committed:
Move defer to beginning of go func

toros11 committed:
Validate that required config variables are set

b0r6 committed:
Fix syntax error

b0r6 committed:
Remove unused import

b0r6 committed:
Fixed a bunch of errors

b0r6 committed:
Fixed some more stuff that broke due to e1341a9

b0r6 committed:
Merge branch 'master' into esxi-driver

b0r6 committed:
Merge branch 'master' into esxi-driver

b0r6 committed:
Power off vm after creation to prevent bug.

b0r6 committed:
Stop: Suspend vm before disconnecting serial

b0r6 committed:
Destroy: Skip trying to power off the vm since it should already be off at this point.

toros11 committed:
Merge branch 'master' into feature-esxi-console

toros11 committed:
Merge remote-tracking branch 'origin/feature-esxi-console' into esxi-driver

toros11 committed:
Move defer call to top of go routine

toros11 committed:
Update function call

toros11 committed:
Receive error and set return code if nil

toros11 committed:
Copy directly to console file descriptor

toros11 committed:
Remove unused import

toros11 committed:
Check for error returns

toros11 committed:
Added console cmd test to esxi acceptance tests

toros11 committed:
Check if toolbox agent is ready before performing command

toros11 committed:
Use wrapper function

toros11 committed:
Remove fmt dependency

toros11 committed:
Remove sleep call

toros11 committed:
Small fix and refactor

Added the function that checks toolbox to the correct type and
renamed it to better reflect the task.

toros11 committed:
Added library that got removed by mistake

b0r6 committed:
Use metadrive to set up network interfaces

b0r6 committed:
Update vendor.json

b0r6 committed:
Merge branch 'esxi-driver'

b0r6 committed:
gofmt

b0r6 committed:
Update vendor.json again

b0r6 committed:
Set type

b0r6 committed:
Remove local metadrive.img after having uploaded it to esxi

b0r6 committed:
Remove img file from datastore folder before removing vm

b0r6 committed:
Set dosfstools as required for executor-esxi

b0r6 committed:
Set hostname

b0r6 committed:
Fixed bug that caused the script to fail when trying to assign more than one interface

Metallion commented on ci/citest/acceptance-test-esxi/tests/01_start_and_terminate_esxi_instance_test.go:
```diff
@@ -0,0 +1,24 @@
+	_, _ = tests.RunCmdWithTimeoutAndReportFail(t, 10, 5, "openvdc", "destroy", instance_id)
+	tests.WaitInstance(t, 5*time.Minute, instance_id, "TERMINATED", nil)
+
+	//TODO: Use Esxi api to make sure that the vm gets properly created and destroyed.
```
Why not do this before merge? I'm afraid this will be forgotten about and might come back to bite us otherwise.

b0r6 replied:
As for the ip wait command: This used to be necessary, but indeed not anymore now after a few changes. I'll remove it.

Metallion reviewed (commented):
Just out of curiosity I was wondering why we're waiting until the instance has an IP before we label it as running. For LXC and KVM we set it to running as soon as the KVM process/LXC container starts.

Metallion commented:
I know @unakatsuo reviewed the code before but since there have been new commits since then, maybe he should have another look?

b0r6 committed:
Remove a few unnecessary commands

b0r6 committed:
Remove unused file

toros11 committed:
Refactor metadata drive

Create metadata drive interface for the driver to implement and
convert the operating methods into reusable functions.

toros11 committed:
Uncomment log message

toros11 committed:
Create metadata image if empty

unakatsuo reviewed (changes requested):
Please import ``github.com/pkg/errors`` and wrap the ``err`` variable returned from std libs. Some lines where receive ``err`` squash the original ``err``.

```
if err := os.MkDir("/....."); err != nil {
  return errors.Wrapf(err, "message %s", v1)
}
```

unakatsuo commented on hypervisor/esxi/esxi.go:
```diff
@@ -227,15 +250,121 @@ func (d *EsxiHypervisorDriver) vmPath() string {
+	default:
+		kp := filepath.Join(keyPath, key)
+		if err := os.MkdirAll(kp, os.ModePerm); err != nil {
+			return errors.Errorf("Unable to create folder: %s", kp)
```
wrap ``err``

unakatsuo commented on hypervisor/esxi/esxi.go:
```diff
@@ -227,15 +250,121 @@ func (d *EsxiHypervisorDriver) vmPath() string {
+func renderData(keyPath string, key string, value interface{}) error {
+	switch value.(type) {
+	case string:
+		ioutil.WriteFile(filepath.Join(keyPath, key), []byte(value.(string)), 0644)
```
```
return errors.Wrap(ioutil.WriteFile(filepath.Join(keyPath, key), []byte(value.(string)), 0644))
```

unakatsuo commented on hypervisor/esxi/esxi.go:
```diff
@@ -194,6 +209,14 @@ func join(separator byte, args ...string) string {
+func runCmd(cmd string, args []string) error {
+	c := exec.Command(cmd, args...)
+	if err := c.Run(); err != nil {
+		return errors.Errorf("failed to execute command :%s %s", cmd, args)
```
Wrap ``err``.

unakatsuo commented on handlers/vm/esxi/esxi.go:
```diff
@@ -0,0 +1,106 @@
+		EsxiImage EsxiImage `json:"esxi_image,omitempty"`
+	}
+	if err := json.Unmarshal(in, &json_template); err != nil {
+		return nil, err
```
Wrap ``err``

unakatsuo commented on handlers/vm/esxi/esxi.go:
```diff
@@ -0,0 +1,106 @@
+	flags.IntVar(&vcpu, "vcpu", int(mdst.MinVcpu), "")
+	flags.IntVar(&mem, "memory_gb", int(mdst.MinMemoryGb), "")
+	if err := flags.Parse(args); err != nil {
+		return err
```
Wrap ``err``.

unakatsuo commented on hypervisor/esxi/esxi_console.go:
```diff
@@ -0,0 +1,185 @@
+	go func () {
+		rOut, wOut, err := os.Pipe()
+		if err != nil {
+			log.Info("failed os.Pipe for stdout")
```
``log.Warn()`` with showing ``err`` contents.

unakatsuo commented on hypervisor/esxi/esxi_console.go:
```diff
@@ -0,0 +1,185 @@
+
+		rErr, wErr, err := os.Pipe()
+		if err != nil {
+			log.Info("failed os.Pipe for stderr")
```
``log.Warn()`` with showing ``err`` contents.

unakatsuo commented on hypervisor/util/serial_connection.go:
```diff
@@ -0,0 +1,78 @@
+				 errc <- err
+				break
+			default:
+				n, err := param.Stdin.Read(b)
```
Missing handle ``io.EOF`` error case returned from ``.Read()``.

https://golang.org/pkg/io/#Reader

unakatsuo commented on hypervisor/util/serial_connection.go:
```diff
@@ -0,0 +1,78 @@
+				_, err = param.Stdout.Write(b[0:n])
+				if err != nil {
+					errc <- errors.Wrap(err, "\nFailed to write to the console stdout buffer\n\n")
+				}
```
Might be good to flush the Stdout buffer?

```
if synchable, ok := param.Stdout.(*os.File); ok {
	if err := synchable.Sync(); err != nil {
		log.Warn("Failed %v to flush: %s", param.Stdout, err)
	}
}
```

toros11 committed:
Improve error handling

toros11 committed:
Separate mount/umount into separate functions

toros11 committed:
Fix typo

toros11 committed:
Output warning

toros11 committed:
make clean return on io.EOF

toros11 committed:
flush param.stdout

toros11 committed:
Merge pull request #200 from axsh/refactor-metadata-drive

Refactor metadata drive

toros11 committed:
Include the error with warning

toros11 committed:
Wrap errors

unakatsuo commented on hypervisor/util/serial_connection.go:
```diff
@@ -28,7 +28,11 @@ func (sc *SerialConnection) AttachSerialConsole(param *hypervisor.ConsoleParam,
 				if err != nil {
-					errc <- errors.Wrap(err, "\nFailed to read from the from the console input buffer\n\n")
+					if err == io.EOF {
+						errc <- errors.Wrap(nil, "\nConsole exited by EOF\n\n")
```
The error message at second argument will never be sent back as error.

https://github.com/pkg/errors/blob/master/errors.go#L181-L183

Print as a log message if you need to catch.

toros11 committed:
Don't wrap error message when err is expected to be nil
